### ISSUE TITLE
Saving model takes too long

### ISSUE THREAD

**Issue by shyamupa** [seomoz/word2gauss#16]
The current implementation does not support multiple epochs for training, so I am running the code as follows (I want to track the performance after each epoch, among other things),

```
with open(opts.corpus, 'r') as corpus:
        for iters in range(opts.iters):
            embed.train(iter_pairs(corpus, vocab), n_workers=opts.threads)
            embed.save(opts.model+".tar.gz."+str(iters), vocab=vocab.id2word, full=True)
```

However, saving the model itself takes too long (>2 minutes). Is there a way to achieve this?

**Comment by shyamupa** [comment]
In fact, I realized the above code is not even training for multiple epochs.

**Comment by matt-peters** [comment]
Yes, the current serialization isn't efficient so if the training time per epoch is short then it will be significant.  This wasn't an issue for us since we trained on lots of data and only serialized at the end.  This is partly due to the tar implementation in python.  There is a faster saving/loading implementation on the `zip` branch you can try (although that branch has gone stale now).  If that is still too slow for your purposes then it's easy to implement a faster serialization since the logic is encapsulated in the `save` and `load` methods.

**Comment by matt-peters** [comment]
The `.train` method is incremental -- it applies gradient descent steps to the current model parameters for each bach in the iterator -- so you can use it to train for many epochs.

Your code isn't training for multiple epochs since the iterator `corpus` is exhausted after reading the file once.

You could achieve it with something like:

``` python
for iter in range(opts.iters):
    with open(opts.corpus, 'r') as corpus:
        embed.train(iter_pairs(corpus, vocab), n_workers=opts.threads))
```

This doesn't shuffle the data between epochs but if your corpus isn't too large you could just load it into memory once then shuffle for each epoch.
